doc/developer/design/20230531_system_privileges.md

```diff
@@ -0,0 +1,86 @@
+# System Privileges
+
+- Associated: [RBAC](20230216_role_based_access_control.md)
+
+## Context
+
+Role based access control comprises a set of features that allow customers to control the
+privileges of individual users. Object privileges control user's access to database objects, they
+are granted with the `GRANT` keyword and revoked with the `REVOKE` keyword. For example,
+`GRANT SELECT ON TABLE t TO r` will grant the ability to read table `t` to role `r`. `GRANT CREATE
+ON SCHEMA s TO r` will grant the ability to create new objects in schema `s` to role `r`. Object
+privileges can be inherited through role membership. Role attributes control the user's ability to
+create (and sometimes modify) top level objects, they are granted and revoked with the `ALTER ROLE`
+keywords. For example, `ALTER ROLE r CREATEDB` will grant the ability to create databases to
+role `r`. `ALTER ROLE r CREATECLUSTER` will grant the ability to create clusters to role `r`. Role
+attributes cannot be inherited through role membership.
+
+Role attributes present some problems as they currently exist:
+
+- It can be confusing that there are two separate privilege primitives that control a role's
+  privilege.
+- Role attributes are not inherited which makes it cumbersome to grant attributes to a large
+  number of roles.
+
+## Goals
+
+- Reduce confusion around privileges.
+- Allow the ability to create databases, clusters, and roles to be inherited through role
+  membership.
+- Maintain PostgreSQL compatibility.
+
+## Overview
+
+We will add a new type of privilege called System Privileges. They will look and act just like
+normal object privileges, but they will grant the abilities that would otherwise be granted through
+role attributes. They will exist in parallel with role attributes.
+
+NOTE: Other databases, such as CockroachDB and Snowflake, have some form of system privileges that
+control similar privileges to role attributes.
+
+## Detailed description
+
+System privileges will be granted with the following syntax:
+`GRANT <privilege-specification> ON SYSTEM TO <role-specification>`
+
+System privileges will be revoked with the following syntax:
+`REVOKE <privilege-specification> ON SYSTEM TO <role-specification>`
+
+System privileges and role attributes will interact in the following ways:
+
+- Granting a system privilege will grant the privilege AND the attribute.
+- Revoking a system privilege will revoke the privilege AND the attribute.
+- Altering a role attribute will affect the role attribute but will not change the system privilege.
```

Comment:
What's an example of altering a role attribute? Is it possible to alter a system privilege?

Reply:
Role attributes always exist in a boolean state. To change the state of a role attribute you must alter the role.

> What's an example of altering a role attribute?
For example `ALTER ROLE joe CREATEDB` sets the `CREATEDB` attribute to `true`, and `ALTER ROLE NOCREATEDB` sets the `CREATEDB` attribute to `false`. Here's a more detailed example:
```
postgres=# CREATE ROLE r1;
CREATE ROLE
postgres=# SELECT rolcreatedb FROM pg_roles WHERE rolname = 'r1';
 rolcreatedb 
-------------
 f
(1 row)

postgres=# ALTER ROLE r1 CREATEDB;
ALTER ROLE
postgres=# SELECT rolcreatedb FROM pg_roles WHERE rolname = 'r1';
 rolcreatedb 
-------------
 t
(1 row)

postgres=# ALTER ROLE r1 NOCREATEDB;
ALTER ROLE
postgres=# SELECT rolcreatedb FROM pg_roles WHERE rolname = 'r1';
 rolcreatedb 
-------------
 f
(1 row)

```

On the other hand, privileges either exist or they don't exist. You create them by granting a privilege via `GRANT ...` and you destroy them via `REVOKE ...`

I think these differences are part of what leads to the confusion between attributes and privileges.

> Is it possible to alter a system privilege?

No, to keep system privileges consistent with the existing privileges you will be able to grant or revoke them but not alter them.

Reply:
Should I add some of the above to the context section?

Reply:
Yes, that would be helpful.

Reply:
Done, I added this towards the end of the Context section.